File: packages/shared-types/src/hours.ts
/**
 * Opening hours are stored as JSON keyed by day (design doc §6.2):
 *   { "mon": "07:00-22:00", "fri": "07:00-14:00,17:00-20:00", "sun": "closed" }
 * A missing day means "unknown", not "closed". Ranges may cross midnight ("20:00-02:00").
 */
export const DAY_KEYS = ['sun', 'mon', 'tue', 'wed', 'thu', 'fri', 'sat'] as const;
export type DayKey = (typeof DAY_KEYS)[number];
export type WeeklyHours = Partial<Record<DayKey, string>>;

export type OpenState = 'open' | 'closed' | 'unknown';

const dayKey = (at: Date): DayKey => DAY_KEYS[at.getDay()] as DayKey;

function toMinutes(hhmm: string): number | null {
  const m = /^(\d{1,2}):(\d{2})$/.exec(hhmm.trim());
  if (!m) return null;
  return Number(m[1]) * 60 + Number(m[2]);
}

function rangeContains(range: string, minutes: number): boolean | null {
  const [start, end] = range.split('-').map(toMinutes);
  if (start == null || end == null) return null;
  if (end <= start) return minutes >= start || minutes < end; // crosses midnight
  return minutes >= start && minutes < end;
}

/** Is a place open at `at`? Uses the device/server's local time, which is Central time on campus. */
export function getOpenState(
  hours: WeeklyHours | null | undefined,
  at: Date = new Date(),
): OpenState {
  if (!hours) return 'unknown';
  const day = hours[dayKey(at)];
  if (day == null || day.trim() === '') return 'unknown';
  if (day.trim().toLowerCase() === 'closed') return 'closed';
  if (day.trim().toLowerCase() === '24h') return 'open';
  const minutes = at.getHours() * 60 + at.getMinutes();
  let parsedAny = false;
  for (const range of day.split(',')) {
    const hit = rangeContains(range, minutes);
    if (hit === null) continue;
    parsedAny = true;
    if (hit) return 'open';
  }
  return parsedAny ? 'closed' : 'unknown';
}

/** Human-readable hours for today, e.g. "7:00 AM – 2:00 PM". */
export function formatTodayHours(
  hours: WeeklyHours | null | undefined,
  at: Date = new Date(),
): string | null {
  const day = hours?.[dayKey(at)];
  if (!day) return null;
  if (day.toLowerCase() === 'closed') return 'Closed today';
  if (day.toLowerCase() === '24h') return 'Open 24 hours';
  return day
    .split(',')
    .map((range) => range.split('-').map(formatClock).join(' – '))
    .join(', ');
}

function formatClock(hhmm: string): string {
  const mins = toMinutes(hhmm);
  if (mins == null) return hhmm;
  const h24 = Math.floor(mins / 60) % 24;
  const m = mins % 60;
  const suffix = h24 < 12 ? 'AM' : 'PM';
  const h12 = h24 % 12 === 0 ? 12 : h24 % 12;
  return `${h12}:${String(m).padStart(2, '0')} ${suffix}`;
}

File: apps/ai-service/app/hours.py
"""Python twin of packages/shared-types/src/hours.ts — keep the two in sync."""

from __future__ import annotations

import re
from datetime import datetime

DAY_KEYS = ["mon", "tue", "wed", "thu", "fri", "sat", "sun"]  # datetime.weekday() order
DAY_NAMES = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
_CLOCK = re.compile(r"^(\d{1,2}):(\d{2})$")


def _minutes(hhmm: str) -> int | None:
    m = _CLOCK.match(hhmm.strip())
    return int(m.group(1)) * 60 + int(m.group(2)) if m else None


def open_state(hours: dict | None, at: datetime) -> str:
    """Returns 'open', 'closed' or 'unknown'. A missing day means unknown, not closed."""
    if not hours:
        return "unknown"
    day = hours.get(DAY_KEYS[at.weekday()])
    if not day or not day.strip():
        return "unknown"
    day = day.strip().lower()
    if day == "closed":
        return "closed"
    if day == "24h":
        return "open"
    now = at.hour * 60 + at.minute
    parsed_any = False
    for rng in day.split(","):
        parts = [_minutes(p) for p in rng.split("-")]
        if len(parts) != 2 or None in parts:
            continue
        parsed_any = True
        start, end = parts
        inside = (now >= start or now < end) if end <= start else (start <= now < end)
        if inside:
            return "open"
    return "closed" if parsed_any else "unknown"


def describe_week(hours: dict | None) -> str:
    if not hours:
        return "hours not recorded"
    return "; ".join(f"{DAY_NAMES[i]} {hours[k]}" for i, k in enumerate(DAY_KEYS) if hours.get(k))
